src/gui/callstatechecker.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2022 by Claudio Cambra <claudio.cambra@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonArray>
+
+#include "callstatechecker.h"
+#include "account.h"
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcCallStateChecker, "nextcloud.gui.callstatechecker", QtInfoMsg)
+
+constexpr int successStatusCode = 200;
+
+CallStateChecker::CallStateChecker(QObject *parent)
+    : QObject(parent)
+{
+    setup();
+}
+
+void CallStateChecker::setup()
+{
+    _notificationTimer.setSingleShot(true);
+    _notificationTimer.setInterval(60 * 1000);
+    connect(&_notificationTimer, &QTimer::timeout, this, &CallStateChecker::slotNotificationTimerElapsed);
+
+    _statusCheckTimer.setInterval(2 * 1000);
+    connect(&_statusCheckTimer, &QTimer::timeout, this, &CallStateChecker::slotStatusCheckTimerElapsed);
+}
+
+QString CallStateChecker::token() const
+{
+    return _token;
+}
+
+void CallStateChecker::setToken(const QString &token)
+{
+    _token = token;
+    Q_EMIT tokenChanged();
+    reset();
+}
+
+AccountState* CallStateChecker::accountState() const
+{
+    return _accountState;
+}
+
+void CallStateChecker::setAccountState(AccountState *state)
+{
+    _accountState = state;
+    Q_EMIT accountStateChanged();
+    reset();
+}
+
+void CallStateChecker::startChecking()
+{
+    qCInfo(lcCallStateChecker, "Starting to check state of call.");
+    _notificationTimer.start();
+    _statusCheckTimer.start();
+}
+
+void CallStateChecker::stopChecking()
+{
+    qCInfo(lcCallStateChecker, "Stopping checking of call state.");
+    _notificationTimer.stop();
+    _statusCheckTimer.stop();
+    _stateCheckJob.clear();
+}
+
+void CallStateChecker::reset()
+{
+    qCInfo(lcCallStateChecker, "Resetting call check");
+    stopChecking();
+    startChecking();
+}
+
+void CallStateChecker::slotNotificationTimerElapsed()
+{
+    qCInfo(lcCallStateChecker, "Notification timer elapsed, stopping call checking");
```

Comment:
maybe we could also add the `_token` in case we would need to analyze any future issue